adafruit_pioasm.py

```diff
@@ -142,9 +142,20 @@ def assemble(text_program):
             #                instr delay dst op src
             assembled.append(0b101_00000_000_00_000)
             assembled[-1] |= MOV_DESTINATIONS.index(instruction[1]) << 5
-            assembled[-1] |= MOV_SOURCES.index(instruction[-1])
-            if len(instruction) > 3:
-                assembled[-1] |= MOV_OPS.index(instruction[-2]) << 3
+            source = instruction[-1]
+            source_split = mov_splitter(source)
+            if len(source_split) == 1:
+                assembled[-1] |= MOV_SOURCES.index(source)
+            else:
+                assembled[-1] |= MOV_SOURCES.index(source_split[1])
+                if source[:1] == "!":
+                    assembled[-1] |= 0x08
+                elif source[:1] == "~":
+                    assembled[-1] |= 0x08
+                elif source[:2] == "::":
+                    assembled[-1] |= 0x10
```

Comment:
```suggestion
                assembled[-1] |= MOV_OPS[source_split[0]]
```

Wouldn't this work?

Reply:
The problem is ! and ~ map to the same 0x08, they are functionally equivalent (tested that with the SDK pioasm). If you go based on array position it would have to become sorta a list of arrays like `[None, ["!", "~"], ::, None]` if that makes sense? I figured making it just the if block is easier to understand and I don't see PIO instructions changing anytime soon.

Reply:
Ah, true! Can you switch the if to use `source_split[0]` then? That way you only split in one place. Doing the split before the `|=` will allow you to still support the three part case (with `~ x`) too.

Reply:
After playing a bit more realized source_split[0] has no value as the operator is the delimiter and gets consumed. But I can just add in the if len > 3 original statement to handle cases like `mov y, ~ x` so it handles the space or non-space versions as before.